File: ch03/src/test/java/jpastart/util/DBUtil.java
package jpastart.util;

import org.dbunit.DatabaseUnitException;
import org.dbunit.database.DatabaseConnection;
import org.dbunit.database.IDatabaseConnection;
import org.dbunit.dataset.DataSetException;
import org.dbunit.dataset.IDataSet;
import org.dbunit.dataset.xml.XmlDataSet;
import org.dbunit.operation.DatabaseOperation;

import java.io.InputStreamReader;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;

public class DBUtil {
    static {
        try {
            Class.forName("org.mariadb.jdbc.Driver");
        } catch (ClassNotFoundException e) {
            throw new RuntimeException(e);
        }
    }

    public static void initTestData() {
        initData("/test-data.xml");
    }

    public static void initData(String dataFile) {
        IDatabaseConnection conn = null;
        try {
            conn = getDatabaseConnection();
            IDataSet data = createDataSet(dataFile);
            DatabaseOperation.CLEAN_INSERT.execute(conn, data);
        } catch (Throwable e) {
            throw new RuntimeException(e);
        } finally {
            close(conn);
        }
    }

    public static IDatabaseConnection getDatabaseConnection() throws SQLException, DatabaseUnitException {
        return new DatabaseConnection(getConnection());
    }

    public static Connection getConnection() throws SQLException {
        return DriverManager.getConnection(
                "jdbc:mariadb://localhost/jpastart?characterEncoding=utf8",
                "jpauser", "jpapass");
    }

    private static IDataSet createDataSet(String resource) throws DataSetException {
        return new XmlDataSet(new InputStreamReader(DBUtil.class.getResourceAsStream(resource)));
    }

    private static void close(IDatabaseConnection conn) {
        if (conn != null) {
            try {
                conn.close();
            } catch (SQLException e) {
                // do nothing
            }
        }
    }

}

File: ch17/src/test/kotlin/jpastart/util/DBUtil.kt
package jpastart.util

import org.dbunit.DatabaseUnitException
import org.dbunit.database.DatabaseConfig
import org.dbunit.database.DatabaseConnection
import org.dbunit.database.IDatabaseConnection
import org.dbunit.dataset.IDataSet
import org.dbunit.dataset.xml.XmlDataSet
import org.dbunit.ext.MariaDBDataTypeFactory
import org.dbunit.operation.DatabaseOperation
import java.io.InputStreamReader
import java.sql.Connection
import java.sql.DriverManager
import java.sql.SQLException

object DBUtil {
    init {
        try {
            Class.forName("org.mariadb.jdbc.Driver")
        } catch (e: ClassNotFoundException) {
            throw RuntimeException(e)
        }
    }

    fun initTestData() {
        initData("/test-data.xml")
    }

    private fun initData(dataFile: String) {
        var conn: IDatabaseConnection? = null
        try {
            conn = databaseConnection
            val data = createDataSet(dataFile)
            DatabaseOperation.CLEAN_INSERT.execute(conn, data)
        } catch (e: Throwable) {
            throw RuntimeException(e)
        } finally {
            close(conn)
        }
    }

    @get:Throws(SQLException::class, DatabaseUnitException::class)
    val databaseConnection: IDatabaseConnection
        get() {
            val conn = DatabaseConnection(connection)
            conn.config.setProperty(DatabaseConfig.PROPERTY_DATATYPE_FACTORY, MariaDBDataTypeFactory())
            return conn
        }

    @get:Throws(SQLException::class)
    val connection: Connection
        get() = DriverManager.getConnection(
            "jdbc:mariadb://localhost/jpastart?characterEncoding=utf8",
            "jpauser", "jpapass"
        )

    private fun createDataSet(resource: String): IDataSet {
        return XmlDataSet(InputStreamReader(DBUtil::class.java.getResourceAsStream(resource)))
    }

    private fun close(conn: IDatabaseConnection?) {
        if (conn != null) {
            try {
                conn.close()
            } catch (e: SQLException) {
                // do nothing
            }
        }
    }


}
